lib/charms/dns_transfer/v0/dns_transfer.py

```diff
@@ -0,0 +1,532 @@
+
+        Returns:
+            Dict containing the representation.
+        """
+        dumped_model = self.model_dump(exclude_unset=True)
+        dumped_data = {}
+        for key, value in dumped_model.items():
+            dumped_data[key] = json.dumps(value, default=str)
+        return dumped_data
+
+    @classmethod
+    def from_relation_data(cls, relation: ops.Relation) -> "DNSTransferProviderData":
+        """Initialize a new instance of the DNSTransferProviderData class from the relation.
+
+        Args:
+            relation: the relation.
+
+        Returns:
+            A DNSTransferProviderData instance.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        try:
+            loaded_data = {}
+            app = cast(ops.Application, relation.app)
+            relation_data = relation.data[app]
+            for key, value in relation_data.items():
+                loaded_data[key] = json.loads(value)
+            return DNSTransferProviderData.model_validate(loaded_data)
+        except json.JSONDecodeError as ex:
+            raise ValueError from ex
+
+
+class DNSTransferRequirerData(pydantic.BaseModel):
+    """List of information for the requirer to manage and transfer.
+
+    Attributes:
+        addresses: IP list of the units composing the provider's deployment.
+    """
+
+    addresses: list[pydantic.IPvAnyAddress]
+
+    # pydantic wants 'self' as first argument
+    @pydantic.field_validator("addresses", mode="before")
+    def deduplicate_addresses(cls, v: Any) -> list:  # noqa: N805 pylint: disable=E0213
+        """Deduplicate addresses.
+
+        Args:
+            v: value
+
+        Returns:
+            list with unique values.
+        """
+        return list(set(v))
+
+    def to_relation_data(self) -> dict[str, str]:
+        """Convert an instance of DNSTransferRequirerData to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        return {"addresses": json.dumps(self.addresses, default=str)}
+
+    @classmethod
+    def from_relation_data(cls, relation: ops.Relation) -> "DNSTransferRequirerData":
+        """Initialize a new instance of the DNSTransferRequirerData class from the relation.
+
+        Args:
+            relation: the relation.
+
+        Returns:
+            A DNSTransferRequirerData instance.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        try:
+            loaded_data = {}
+            app = cast(ops.Application, relation.app)
+            relation_data = relation.data[app]
+            for key, value in relation_data.items():
+                loaded_data[key] = json.loads(value)
+            return DNSTransferRequirerData.model_validate(loaded_data)
+        except json.JSONDecodeError as ex:
+            raise ValueError from ex
+
+
+class DNSTransferRequirerDataAvailableEvent(ops.RelationEvent):
```

Comment:
So the added value compared the usual RelationChangedEvent is that this way you add a shortcut function to access the relation data ?  
I'm not sure that I value this enough for the added code but it's a matter of opinion

Reply:
Good point. As we discussed previously in our 1:1, this can be removed.